src/base.c

```diff
@@ -491,6 +539,63 @@ base_alloc_edata(tsdn_t *tsdn, base_t *base) {
 	return edata;
 }
 
+static inline void
+b0_alloc_header_size(size_t *header_size, size_t *alignment) {
+	*alignment = QUANTUM;
+	*header_size = QUANTUM > sizeof(edata_t *) ? QUANTUM :
+	    sizeof(edata_t *);
+}
+
+void *
+b0_alloc_tcache_stack(tsdn_t *tsdn, size_t stack_size) {
+	base_t *base = b0get();
+	edata_t *edata = base_alloc_base_edata(tsdn, base);
+	if (edata == NULL) {
+		return NULL;
+	}
+
+	/*
+	 * Reserve room for the header so that edata can be retrieved on dalloc.
+	 * Bump up to usize to improve reusability -- otherwise the freed stacks
+	 * will be put back into the previous size class.
+	 */
+	size_t esn, alignment, header_size;
+	b0_alloc_header_size(&header_size, &alignment);
+
+	size_t alloc_size = sz_s2u(stack_size + header_size);
+	void *addr = base_alloc_impl(tsdn, base, alloc_size, alignment, &esn);
+	if (addr == NULL) {
+		edata_avail_insert(&base->edata_avail, edata);
+		return NULL;
+	}
+
+	edata_binit(edata, addr, alloc_size, esn, true /* is_reused */);
+	*(edata_t **)addr = edata;
+
+	return (byte_t *)addr + header_size;
```

Comment:
I am confused here. `edata` here is a allocated through `base_alloc_base_edata` while `addr` is allocated from `base_alloc_impl`. The two allocations are not supposed to be continuous, right? Why are we returning the `edata + header_size` here? 

Also `addr` is allocated a memory with `alloc_size`, which should include the size of an `edata_t`. Why do we need an extra `header_size` to be allocated there since we already have an `edata_t`?

Reply:
I should add more comments explaining the design:

Yes they are two separate allocations. Similar to the regular memory and their metadata, the edata_t tracks the info about the allocated memory / extent. One key difference is that, we cannot use rtree lookups to find the metadata on deallocation, and we need a way to do so.

That's what's happening here: we are adding a tiny header (effectively only a pointer: note that alloc_size includes a `sizeof(edata_t *)`, not the actual edata_t), which is stored right before the allocated addr -- then on deallocation, we can simply offset and retrieve the pointer, which finds us the edata_t (that's the beginning part of `b0_dalloc_tcache_stack`).